pysmt/optimization/optimizer.py

```diff
@@ -0,0 +1,703 @@
+            bool: True if progress can still be made (i.e., constraints are satisfiable),
+                  False otherwise.
+
+        Raises:
+            NotImplementedError: If the method is not implemented in the subclass.
+        """
+        raise NotImplementedError()
+
+    def _pareto_block_model(self, client_data, objs):
+        """
+        Blocks the current model in the Pareto optimization process.
+
+        This method constructs a disjunction of constraints to block the current model,
+        ensuring that the solver does not return the same solution in subsequent iterations.
+
+        Parameters:
+            client_data (list): A list of constraints accumulated during the optimization process.
+            objs (list): A list of `OptPareto` objects representing the current optimization objectives.
+
+        Returns:
+            None
+
+        Raises:
+            NotImplementedError: If the method is not implemented in the subclass.
+        """
+        raise NotImplementedError()
+
+
+class SUAOptimizerMixin(ExternalOptimizerMixin):
+    """Optimizer mixin using solving under assumptions"""
+
+    def _optimization_check_progress(self, client_data, formula, strategy, extra_assumption = None):
+        assum = extra_assumption if extra_assumption is not None else []
+        if formula is not None:
+            assum = assum + [formula]
+        rt = self.solve(assumptions = assum)
+        return rt
+
+    def _lexicographic_opt(self, client_data, current_goal, strategy):
+        temp = self._optimize(current_goal, strategy, extra_assumption = client_data)
+        if temp is not None:
+            model, val = temp
+        else:
+            return None
+
+        client_data.append(Equals(current_goal.term(), val))
+        return model, val
+
+    def _pareto_check_progress(self, client_data, objs):
+        mgr = self.environment.formula_manager
+        k = []
+        if objs[0].val is not None:
+            k = [obj.get_constraint(False) for obj in objs]
+            k.append(mgr.Or(obj.get_constraint(True) for obj in objs ))
+        return not self.solve(assumptions=client_data + k)
+
+    def _pareto_block_model(self, client_data, objs):
+        mgr = self.environment.formula_manager
+        client_data.append(mgr.Or(obj.get_constraint(True) for obj in objs))
+
+
+class IncrementalOptimizerMixin(ExternalOptimizerMixin):
+    """Optimizer mixin using the incremental capabilites of the solver"""
+
+    def _optimization_check_progress(self, client_data, formula, strategy, extra_assumption = None):
+        if strategy == 'linear':
+            if formula is not None:
+                self.add_assertion(formula)
+            rt = self.solve()
+        elif strategy == 'binary':
+            self.push()
+            if formula is not None:
+                self.add_assertion(formula)
+            rt = self.solve()
+            # TODO for example in z3 the get_model must be called before the pop.
```

Comment:
I think this is a design bug: after the pop() we are NOT guaranteed to have access to the model, so it would be better to get it before the pop() and return it to the caller (also in the pareto case)